### ISSUE TITLE
"Manage events" always stores NULL for 10:00 time limits in DB

### ISSUE THREAD

**Issue by SAuroux** [thewca/worldcubeassociation.org#5460]
I am creating this issue after already discussing the matter with @viroulep. This bug/issue is interfering with a current WRT project, so having a solution for it soon would be highly appreciated.

One clarification upfront: I am aware that NULL values for rounds.time_limit in the database are represented as 10:00 on the website and that the TimeLimit class also defaults to 10 minutes in this case. This is certainly related, but **not** the subject of this issue.

**Describe the bug**

When creating a new round on the "Manage events" page or when changing a time limit to 10:00, NULL is stored to rounds.time_limit instead of '{"centiseconds":60000,"cumulativeRoundIds":[]}' when saving. 

**Expected behavior**

- Generating a new round with time limit 10:00 or changing the time limit to 10:00 should results in a value of '{"centiseconds":60000,"cumulativeRoundIds":[]}' in the database.
- The website should not modify rounds.time_limit when no change was made, certainly not from '{"centiseconds":60000,"cumulativeRoundIds":[]}' to NULL, but also not from NULL to '{"centiseconds":60000,"cumulativeRoundIds":[]}' (as NULL is still a preferred default value for competitions before 2013).

**To Reproduce**
Scenarios I tried on staging:

1. Adding new round for some event of a competition and save without changing the fdefault time limit of 10:00 -> rounds.time_limit is NULL.
2. Changing the time limit to 9:00 for some event of a competition and save -> rounds.time_limit is '{"centiseconds":54000,"cumulativeRoundIds":[]}'. Then change the time limit back to 10:00 and save -> rounds.time_limit is back to NULL.
3. Manually change rounds.time_limit from NULL to '{"centiseconds":60000,"cumulativeRoundIds":[]}' for some event of a competition in the database, go to the competition's "Manage events" page, make a completely unrelated change (e.g. add a new event) and save -> rounds.time_limit is back to NULL.

**Comment by viroulep** [comment]
> This is certainly related, but not the subject of this issue.

Turns out it is actually, as this issue can't be solve without addressing the "default" time limit for null values.

It's much more tricky than I thought...
This issue comes from how rails handle serialization: when we're about to serialize an attribute, rails checks if that's the default value (constructed from a call to `Obj.load(nil)`), and if yes, just store null, otherwise store the serialized object.
While it makes sense, for us it's a bit annoying because so far we've always considered that a round would have a time limit, and therefore picking 10 minutes as the default made sense.

However given the WRT would like to use null values for unknown time limit (for competitions < 2013 for instance), we need to make sure that:
  - constructing a time limit from "nothing" would result in a 10 minutes time limit (for instance when adding a round)
  - constructing a time limit from null in the db should be interpreted as an undefined time limit

I've quickly looked at supporting null time limit in the codebase but it's painful, so I've come up with the proposal in #5463.
It considers a time limit of 0 seconds as the time limit constructed from a null value (and therefore saving a 10 minutes time limit would actually result in writing the appropriate value to the db), and renders it as an empty string on the "Events" tab of the competition page.

@SAuroux: for this to work appropriately, we would need to update the currently null value in the db in some way:
  - we could update all null value to the actual value for a 10 minutes time limit (I've included a migration for that)
  - we could update only null values for rounds at competitions after a given date
 
If we update all, the WRT could then "cherry-pick" the rounds/competitions which should actually be null in the db (and would be represented as empty strings on the events page).

**Comment by SAuroux** [comment]
Thanks for your work @viroulep, all of this sounds good to me. The initial migration should update only the NULL values for competitions with year >= 2013 in my opinion.

**Comment by viroulep** [comment]
What about the rounds <2013: are they any which are actually correctly setup with a 10 minutes time-limit?
Or will the WRT just do a pass on all them?

**Comment by SAuroux** [comment]
Certainly many of them are, but as before 2013 were allowed to surpass time limits at the discretion of the delegate (so in practice people were essentially never stopped at the time limit), the time limit before 2013 would be a different sort of time limit than the one >=2013.

So from validation perspective, I think that keeping NULL for rounds < 2013 is more appropriate, because we would receive many validation errors otherwise that are not valid for these rounds.

Ideally, as the time limit was still 10:00 by default since 2007 already, we would still display NULL as 10:00 on the website for 2007 >= round < 2013, but this might be difficult to realize in a clear manner, right?

**Comment by viroulep** [comment]
> Ideally, as the time limit was still 10:00 by default since 2007 already, we would still display NULL as 10:00 on the website for 2007 >= round < 2013, but this might be difficult to realize in a clear manner, right?

Right, this would really be a pain to implement cleanly because we would have to associate two different representations to the same value.

> Certainly many of them are, but as before 2013 were allowed to surpass time limits at the discretion of the delegate (so in practice people were essentially never stopped at the time limit), the time limit before 2013 would be a different sort of time limit than the one >=2013.

I forgot that detail!
Then yes it definitely makes sense to let them undefined.
The results validator would have reported an error for them though, as it would be detected as present and equal to 0; I just update the PR to ignore time limit validation if it's undef, and [emit a warning](https://github.com/thewca/worldcubeassociation.org/pull/5463/files#diff-a59f1336c36da875ef2cd06762ff812cR18) that it should occur only for <2013 competitions.
